### ISSUE TITLE
Any plans for Cursor Support?

### ISSUE THREAD

**Issue by tlhunter** [tgriesser/knex#656]
Is there any plan to add support for PostgreSQL's Cursors?
http://www.postgresql.org/docs/9.2/static/plpgsql-cursors.html

It looks like MySQL supports them as well but only within stored procedures:
http://dev.mysql.com/doc/refman/5.0/en/cursors.html

**Comment by bendrucker** [comment]
There are a bunch of open issues about Postgres-specific features. To be perfectly honest, I don't think cursor support is particularly high on that list.

**Comment by tgriesser** [comment]
I'll take a look into this

**Comment by calvinmetcalf** [comment]
@tlhunter fyi when you stream a query out of postgres is uses cursors

**Comment by jonstuebe** [comment]
Looks like this might be helpful in extending `pg`:

https://www.npmjs.com/package/pg-cursor

**Comment by tomjrees** [comment]
I was googling around for a solution and this seems to be the top result. I did manage to get `pg-cursor` to play nicely with Knex, although it's a bit of a hack to access a raw PG connection. Pasting it here because this might be useful to somebody.

(My use-case: Streaming from a large table of 500k+ rows to a CSV export feature)

```typescript
import pg from 'pg'
import Cursor from 'pg-cursor'
import { knex } from './db/knex'

async function cursorDemo() {
  // Access the underlying Knex postgres connection:
  // (See: knex/lib/dialects/postgres/index.js)
  const knexPostgresClient = (knex as any).connection().client

  let client: pg.Client | null = null
  try {
    // Probably upsetting to the knex authors:
    client = (await knexPostgresClient.acquireRawConnection()) as pg.Client

    // 500k+ rows
    const text = `SELECT * FROM view_traffic_90`
    const cursor = client.query(new Cursor(text, []))

    while (true) {
      const rows: unknown[] = await new Promise((resolve, reject) => {
        // Read 10k rows at a time; these will be sent to a Readable stream
        cursor.read(10000, (err, rows) => {
          if (err) {
            reject(err)
          } else {
            resolve(rows)
          }
        })
      })
      /* ... do something with rows ... */
      console.log(`Read ${rows.length} rows`)
      if (rows.length === 0) {
        break
      }
    }
  } finally {
    if (client) {
      await knexPostgresClient.destroyRawConnection(client)
    }
  }
}
```

**Comment by tomjrees** [comment]
After looking at the underlying code some more: knex's `.stream()` method uses `pg-query-parser` under the hood, which is a wrapper around `pg-cursor`. So this feature is fully implemented inside knex. The issue should probably be closed.

You can achieve the same as the above code by just calling `knex.stream({ highWaterMark: 10000 })` and piping the results eg. to CSV.stringify.

Documentation for this is pretty hard to find though.

**Comment by pthieu** [comment]
@teajaymars do you know if highWaterMark is the standard node.js stream unit of bytes or is it number of rows?

**Comment by mercmobily** [comment]
As mentioned above, Knex actually has full cursor support for PostgreSQL today, though it's implemented in a way that might not be immediately obvious from the documentation. The `.stream()` method in Knex uses PostgreSQL cursors under the hood through the `pg-query-stream` library, providing efficient memory usage for large result sets without requiring manual cursor management.

Looking at the implementation in lib/dialects/postgres/index.js:187-212, when you call `.stream()` on a PostgreSQL query, Knex creates a server-side cursor through pg-query-stream. This module, as confirmed in its documentation, "uses a cursor on the server so it efficiently keeps only a low number of rows in memory." This is exactly the functionality you were requesting.

Here's how you can use cursor functionality in Knex today:

```javascript
// Basic streaming with default cursor settings
const stream = knex('large_table').stream();
stream.on('data', (row) => {
  // Process each row as it comes from the cursor
});

// Control cursor batch size with highWaterMark
const stream = knex('large_table').stream({ highWaterMark: 100 });
// This fetches 100 rows at a time from the cursor

// Use as async iterator (modern approach)
const stream = knex('large_table').stream();
for await (const row of stream) {
  // Process rows one by one
}
```

The highWaterMark option controls how many rows the cursor fetches at a time, giving you fine control over memory usage versus network round trips. This effectively provides the same functionality as PostgreSQL's DECLARE CURSOR and FETCH commands, but with a more idiomatic Node.js streaming interface.

Regarding Tim's comment about accessing raw cursors directly through pg-cursor, that's also possible if you need lower-level control. As shown in the discussion by teajaymars, you can access the underlying PostgreSQL connection and use pg-cursor directly for specialized use cases:

```javascript
const client = await knex.client.acquireRawConnection();
const Cursor = require('pg-cursor');
const cursor = client.query(new Cursor('SELECT * FROM huge_table', []));
// Manual cursor control...
```

However, for most use cases, the built-in `.stream()` method provides all the cursor functionality needed without the complexity of manual cursor management.

The confusion around this feature stems from a documentation gap rather than missing functionality. The cursor support is there but isn't explicitly documented as "cursor support" - it's presented as "streaming" even though it uses cursors internally. This has led to issues like #2562 where users struggle to get streaming working because they don't realise they need to install pg-query-stream separately.

To summarise: Knex has had PostgreSQL cursor support since at least version 0.8.0 (when streaming was introduced), implemented through the `.stream()` interface. The functionality you requested in 2015 has been available all along, just not explicitly labelled as cursor support in the documentation.

Improving the documentation will resolve this ticket.

**Related Issues:**
- #2562 - Users experiencing issues with streaming when pg-query-stream isn't installed
- #1836 - Stream memory issues that were resolved by proper cursor usage
- #5654 - Recent streaming improvements for large datasets
- #870 - Documentation improvements for streaming behaviour
